### DISCUSSION TITLE
[Proposal]: Error handling or Result<T, E>

### DISCUSSION THREAD

**Discussion by molor824** [dotnet/csharplang#5294] [Language Ideas]
# Error handling or Result<T, E>

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
Return Result<T, E> and make programmer handle the error instead of throwing exception

## Motivation
Instead of throwing exception in the middle of the process which could cause confusion on wheter if its your error or the library error, making the programmer handle the error allows for the programmer to understand the error more easily.

## Detailed design
Simular to rust, `Result<T, E>` will have many ways to handle the error, either by unwrapping which panics if the Result is Error, or through if/switch statement. Unwrapping could possibly be a `.` operator instead of function to allow programmers to completely ignore it. For example `SomeFunc().foo` instead of `SomeFunc().unwrap().foo`

## Drawbacks
Might remove the purpose of exception

## Alternatives
Could return nullable value but then you'll only get null exception when used. `Result<T, E>` throws `E` as its exception making it more clear to understand

## Design meetings
The idea was from rust programming language's error handling https://doc.rust-lang.org/book/ch09-00-error-handling.html

**Comment by svick** [comment]
For a new language, I think the choice between exceptions and something like `Result<T, E>` is a good topic for discussion, since both have their pros and cons and neither choice is clearly better.

For a decades old language, I think there is no choice: the only reasonable approach is to stick with the error-handling strategy you already have. That's because retrofitting a new strategy onto an existing language would lead to a lot of friction (e.g. what happens to all the existing code that uses exceptions?). And if you add the cons of the new approach to that, they significantly outweigh its pros.

**Comment by gulshan** [comment]
Some of my thoughts on the this- #2734

**Comment by HaloFour** [comment]
This should be moved to a discussion.

See: https://github.com/dotnet/csharplang/issues/113

Also, in my opinion, while I do think that there's value in the language adding support for discriminated unions that would make it easier to work with `Option`/`Maybe`/`Result`/`Try`/etc. types, I believe that it is _exceptionally_ unlikely (pun intended) for the [Runtime](https://github.com/dotnet/runtime/) team to consider splitting the ecosystem and introducing a complete new set of BCL APIs that would favor DUs (or anything else) over the existing idiomatic C# approaches of throwing exceptions or following the `Try*` pattern.  Those expecting that the BCL will all of a sudden adopt functional idioms are likely to be very disappointed.

**Reply by DavidArno** [reply]
> Those expecting that the BCL will all of a sudden adopt functional idioms are likely to be very disappointed.

This is true. The BCL's insistence on looking at the implementation details of tuples, thinking in an imperative-first programming way and therefore adopting PascalCase for tuple labels being a classic example of how they resist functional idioms. The key is is to accept that's what they will do and move on. It took me a while, but I got there in the end. 😁 

When DU's land, it'll be possible to adopt functional style try patterns in one's own code, but that will always sit atop an old style BCL. But F# copes just fine with that and so can a more functional C#.

**Reply by julealgon** [reply]
I can see how most APIs will not leverage DUs, but wouldn't it be an overall win to adopt them for the `TryX` pattern methods? This would remove one of the last few cases where `out` parameters are used and lead to cleaner patterns.

Similar API shifts happened when `IEnumerable` was introduced, then when `async` was introduced, and then when `Span<T>/Memory<T>` were introduced. Admittedly, all of these did bring in new functionality.

**Reply by HaloFour** [reply]
@julealgon 

> Admittedly, all of these did bring in new functionality.

And had other obvious benefits, not just for stylistic preferences.  The runtime team had already rejected proposals to add overloads to the `TryX` methods to return nullable value types and I doubt that DUs will do much in the way to change the collective opinion on that subject.

I'd like to be wrong here and if the runtime team did decide that they wanted to add overloads all over the place to support DUs instead of `out` parameters I'd give that effort a big 👍.

**Reply by julealgon** [reply]
I guess a more apt comparison would be with the value tuple-enhanced overloads, for example:
- https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.zip?view=net-5.0#System_Linq_Enumerable_Zip__2_System_Collections_Generic_IEnumerable___0__System_Collections_Generic_IEnumerable___1__

This one is basically syntactic sugar and doesn't bring in new functionality by itself: it just leverages value tuples for the nicer syntax when working with the common scenario of zipping 2 sequences.

While I don't expect widespread support too, I hope that at there is at least _some_ nice integration of DUs like this when it becomes available, in places where the readability/usability benefits are worth it.

**Comment by erik-kallen** [comment]
This does not really need language support, but it could use discriminated unions, which would be my favorite language feature.
